src/transformers/generation/utils.py

```diff
@@ -1778,6 +1798,369 @@ def typeerror():
 
         return result
 
+    def _dola_decoding(
```

Comment:
This method is way too long - it needs to be broken up into smaller functions

Reply:
I moved the core part of the dola decoding: selecting layers and contrasting the logits, into a new function `_dola_select_contrast()`. It reduced the length of `_dola_decoding()` by a lot. 

Let me know if this change is enough! I can also make other changes too! 🤗